compiler/rustc_borrowck/src/type_check/free_region_relations.rs

```diff
@@ -276,6 +277,25 @@ impl<'tcx> UniversalRegionRelationsBuilder<'_, 'tcx> {
             })
             .collect();
 
+        // Add implied bounds from impl header.
+        if let DefKind::AssocFn = tcx.def_kind(defining_ty_def_id) {
```

Comment:
TODO: this should handle `DefKind::AssocConst` as well, but this is blocked on #104763. So we should add a FIXME here and a check-fail test for an attempted exploitation of the unsoundness.